src/main/java/org/apache/maven/plugins/toolchain/jdk/ToolchainDiscoverer.java

```diff
@@ -0,0 +1,433 @@
+                            // Remove stale entries
+                            .filter(tc -> {
+                                // If the bin/java executable is not available anymore, remove this TC
+                                if (!hasJavaC(getJdkHome(tc))) {
+                                    cacheModified = false;
+                                    return false;
+                                } else {
+                                    return true;
+                                }
+                            })
+                            .collect(Collectors.toConcurrentMap(this::getJdkHome, Function.identity()));
+                }
+            }
+        } catch (IOException | XmlPullParserException e) {
+            log.warn("Error reading toolchains cache: " + e);
+        }
+    }
+
+    private void writeCache() {
+        try {
+            Path cacheFile = getCacheFile();
+            Files.createDirectories(cacheFile.getParent());
+            try (Writer w = Files.newBufferedWriter(cacheFile)) {
+                PersistedToolchains pt = new PersistedToolchains();
+                pt.setToolchains(cache.values().stream()
+                        .map(tc -> {
+                            ToolchainModel model = tc.clone();
+                            model.getProvides().remove(CURRENT);
+                            model.getProvides().remove(ENV);
+                            return model;
+                        })
+                        .sorted(version().thenComparing(vendor()))
+                        .collect(Collectors.toList()));
+                new MavenToolchainsXpp3Writer().write(w, pt);
+            }
+        } catch (IOException e) {
+            log.warn("Error writing toolchains cache: " + e);
+        }
+    }
+
+    private static Path getCacheFile() {
+        return Paths.get(System.getProperty(USER_HOME)).resolve(DISCOVERED_TOOLCHAINS_CACHE_XML);
+    }
+
+    private Path getJdkHome(ToolchainModel toolchain) {
+        Xpp3Dom dom = (Xpp3Dom) toolchain.getConfiguration();
+        Xpp3Dom javahome = dom != null ? dom.getChild(JDK_HOME) : null;
+        String jdk = javahome != null ? javahome.getValue() : null;
+        return Paths.get(Objects.requireNonNull(jdk));
+    }
+
+    ToolchainModel getToolchainModel(Path jdk) {
+        log.debug("Computing model for " + jdk);
+
+        ToolchainModel model = cache.get(jdk);
+        if (model == null) {
+            model = doGetToolchainModel(jdk);
+            cache.put(jdk, model);
+            cacheModified = true;
+        }
+
+        return model;
+    }
+
+    ToolchainModel doGetToolchainModel(Path jdk) {
+        Path java = jdk.resolve("bin/java");
+        if (!java.toFile().canExecute()) {
+            java = jdk.resolve("bin/java.exe");
+            if (!java.toFile().canExecute()) {
+                log.debug("JDK toolchain discovered at " + jdk + " will be ignored: unable to find java executable");
```

Comment:
Not quite accurate. It found java. It just can't execute it.

Reply:
I've cleaned the messages a bit.